src/main/java/com/direwolf20/buildinggadgets/common/schema/BGTemplate.java

```diff
@@ -0,0 +1,68 @@
+package com.direwolf20.buildinggadgets.common.schema;
+
+import com.google.common.collect.AbstractIterator;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap.Entry;
+import it.unimi.dsi.fastutil.objects.ObjectIterator;
+import net.minecraft.block.BlockState;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.util.math.BlockPos.Mutable;
+import net.minecraft.util.math.Vec3i;
+
+import java.util.Iterator;
+
+public final class BGTemplate implements Iterable<MutableBlockData> {
+    private static final int BYTE_MASK_12BIT = 0xF_FF;
+    private static final int BYTE_MASK_14BIT = 0x3F_FF;
+    private final Long2ObjectMap<BlockState> states;
+    private final Long2ObjectMap<CompoundNBT> nbtData;
+    private final BoundingBox bounds;
+    private final String author;
+    private final String name;
+
+    private BGTemplate(Long2ObjectMap<BlockState> states, Long2ObjectMap<CompoundNBT> nbtData, BoundingBox bounds,
+                      String author, String name) {
+        this.states = states;
+        this.nbtData = nbtData;
+        this.bounds = bounds;
+        this.author = author;
+        this.name = name;
+    }
+
+    private long vecToLong(Vec3i vec) {
```

Comment:
Just use BlockPos, we're dealing with blocks and the BlockPos class has a Static#fromLong and a public#toLong built into it.

Reply:
I did see these Methods and decided against them for today as I kind of couldn't see what exact bit masks they are using. I intended to calculate the log2 and check if I can use them (Why the hack doesn't MOJANG just hardcode how many bits they want to allocate to each dimension?!?) when I come to the actual serialisation.

Reply:
We don't need to know the exact bit mask as they provide both methods...

Reply:
We need to know the technical limits for a Template's size which are defined by this. Furthermore I thought we wanted to have 14 Bits for x and z and 12 for y, therefore I thought I'd rather make sure we have exactly that size.

Reply:
So I just checked: MC uses all 64 long bits, therefore I can't use that for serialisation, as I'd like to also pack the state id into the long. Therefore I'll already need this Method and may as well use the same format as map keys as I use for serialisation in order to save myself some conversions - I'm going to leave this and add a comment explaining that.